src/Hpack/Config.hs

```diff
@@ -691,37 +729,55 @@ toLibrary dir name globalOptions library = traverse fromLibrarySection sect >>=
+    fromLibrarySectionPlain LibrarySection{..} = Library librarySectionExposed (fromMaybeList librarySectionExposedModules) (fromMaybeList librarySectionOtherModules) (fromMaybeList librarySectionReexportedModules)
+
+toExecutables :: FilePath -> String -> Section Empty global -> [(String, Section ExecutableSection ExecutableSection)] -> IO ([String], [Section Executable Executable])
 toExecutables dir packageName_ globalOptions executables = do
   result <- mapM toExecutable sections >>= mapM (expandForeignSources dir)
   let (warnings, xs) = unzip result
   return (concat warnings, xs)
   where
-    sections :: [(String, Section ExecutableSection)]
-    sections = map (fmap $ mergeSections globalOptions) executables
+    sections :: [(String, Section ExecutableSection ExecutableSection)]
+    sections = map (fmap $ mergeSections emptyExecutable globalOptions) executables
+
+    emptyExecutable :: ExecutableSection
+    emptyExecutable = ExecutableSection Nothing Nothing
 
-    toExecutable :: (String, Section ExecutableSection) -> IO (Section Executable)
+    toExecutable :: (String, Section ExecutableSection ExecutableSection) -> IO (Section Executable Executable)
     toExecutable (name, sect@Section{..}) = do
       (executable, ghcOptions) <- fromExecutableSection sectionData
-      return (sect {sectionData = executable, sectionGhcOptions = sectionGhcOptions ++ ghcOptions})
+      return (sect
+        { sectionData = executable
+        , sectionGhcOptions = sectionGhcOptions ++ ghcOptions
+        , sectionConditionals = (fmap . fmap) fromExecutableSectionNamed sectionConditionals
+        })
       where
         fromExecutableSection :: ExecutableSection -> IO (Executable, [GhcOption])
         fromExecutableSection ExecutableSection{..} = do
           modules <- maybe inferModules (return . fromList) executableSectionOtherModules
-          return (Executable name mainSrcFile modules, ghcOptions)
+          return (Executable (Just name) mainSrcFile modules, ghcOptions)
           where
             inferModules :: IO [String]
             inferModules = filterMain . (++ [pathsModule]) . concat <$> mapM (getModules dir) sectionSourceDirs
 
             pathsModule = pathsModuleFromPackageName packageName_
 
             filterMain :: [String] -> [String]
-            filterMain = maybe id (filter . (/=)) (toModule $ splitDirectories executableSectionMain)
+            filterMain = maybe (const []) (\m -> maybe id (filter . (/=)) (toModule $ splitDirectories m)) executableSectionMain
+
+            (mainSrcFile, ghcOptions) = case executableSectionMain of
+              Nothing -> (Nothing, [])
+              Just m -> let (f, o) = parseMain m in (Just f, o)
 
-            (mainSrcFile, ghcOptions) = parseMain executableSectionMain
+        fromExecutableSectionNamed :: ExecutableSection -> Executable
+        fromExecutableSectionNamed (ExecutableSection main other) = Executable Nothing main (fromMaybeList other)
```

Comment:
I actually infer `other-modules` inside conditional too now (provided the user specified any `source-dirs`).  Any reason we wouldn't want that?